rs/ic_os/manifest_tool/src/components_parser.rs

```diff
@@ -0,0 +1,127 @@
+use anyhow::{Context, Result};
+use regex::Regex;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+pub const COMPONENTS_PATH: &str = "ic-os/components";
+
+#[derive(Debug)]
+pub struct IcosManifest {
+    pub setupos: Manifest,
+    pub hostos: Manifest,
+    pub guestos: Manifest,
+    pub boundary_guestos: Manifest,
+}
+
+impl IcosManifest {
+    pub fn new(
+        guestos_components: Manifest,
+        hostos_components: Manifest,
+        setupos_components: Manifest,
+        boundary_guestos_components: Manifest,
+    ) -> Self {
+        IcosManifest {
+            guestos: guestos_components,
+            hostos: hostos_components,
+            setupos: setupos_components,
+            boundary_guestos: boundary_guestos_components,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Manifest {
+    pub manifest: Vec<ManifestEntry>,
+}
+
+impl Manifest {
+    pub fn new() -> Self {
+        Manifest {
+            manifest: Vec::new(),
+        }
+    }
+
+    pub fn add_entry(&mut self, component: ManifestEntry) {
+        self.manifest.push(component);
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ManifestEntry {
+    pub source: PathBuf,
+    #[allow(dead_code)]
+    pub destination: PathBuf,
+}
+
+impl ManifestEntry {
+    pub fn new(source: PathBuf, destination: PathBuf) -> Self {
+        ManifestEntry {
+            source,
+            destination,
+        }
+    }
+}
+
+fn get_manifest(manifest_path: &Path, components_path: &Path) -> Result<Manifest> {
```

Comment:
This pattern looks pretty similar to [`parse`](https://doc.rust-lang.org/std/primitive.str.html#method.parse) and [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html).

Maybe we could split up the part that reads the file from the part that builds the `Manifest`, so that we can easily build on this later. For example, we could be ready to roll if we swap out the `Regex` impl with something from the `serde` family.

Reply:
Agreed! Updated!